### PR TITLE
Utilities: Introduce the ldd utility

### PR DESCRIPTION
**Pull request by supercomputer7**
This utility lets a user to figure out what are the dependency libraries
for an ELF dynamic object, whether it's a dynamically loaded executable
or dynamically loaded library.

cc @ADKaster

### CODE REVIEW

**Comment by nico** [comment]
Is there a way to print this from just statically looking at the elf headers, without actually mapping the binary?

**Comment by supercomputer7** [comment]
> Is there a way to print this from just statically looking at the elf headers, without actually mapping the binary?

Not that I know of with the current code in `LibELF`, we need to map the binary and find the dynamic section of it.
We could probably add code to try to statically find it but it feels like it's beyond the scope of this PR if there's already a method in `LibELF` to do that.
Also, from the linux manual page about `ldd`:
Thus, you should never employ ldd on an untrusted executable,

> Security
>        Be aware that in some circumstances (e.g., where the program
>        specifies an ELF interpreter other than ld-linux.so), some
>        versions of ldd may attempt to obtain the dependency information
>        by attempting to directly execute the program, which may lead to
>        the execution of whatever code is defined in the program's ELF
>        interpreter, and perhaps to execution of the program itself.  (In
>        glibc versions before 2.27, the upstream ldd implementation did
>        this for example, although most distributions provided a modified
>        version that did not.)
> 
>        Thus, you should never employ ldd on an untrusted executable,
>        since this may result in the execution of arbitrary code.  A
>        safer alternative when dealing with untrusted executables is:
> 
>            $ objdump -p /path/to/program | grep NEEDED
> 
>        Note, however, that this alternative shows only the direct
>        dependencies of the executable, while ldd shows the entire
>        dependency tree of the executable.

But we don't have objdump as far as I can see, so we are basically left with building our own `ldd` implementation, which is also safe to use on untrusted binaries because we never use the specified interpreter to find the dependencies.

**Comment by ADKaster** [comment]
In theory we have the machinery to do it without *linking* the library. However it looks like ELF::DynamicObject expects that the library is created from the DynamicLoader::map method, which does require at least loading the program headers (and all loadable sections :( ) somewhere in memory. Which is the solution presented in this PR. 

I think we could definitely do less work with some LibELF re-architecting though. We shouldn't have to actually mmap all of the loadable segments into memory in order to check the dynamic section for DT_NEEDED entries.

We would just need to, for each library:
- Open it up and find the dynamic section
- Grab all the DT_NEEDED entries
- Looking at LD_LIBRARY_PATH, builtin library paths, and RPATH of the current object, find the path to all the DT_NEEDED objects
- Keep doing that recursively.

Doing that in a way that will stay in lock step with the way dependencies are resolved in the loader itself sounds hard though. Any differences in  how they resolve libraries could cause some serious headaches when checking ldd vs running a binary.

**Comment by supercomputer7** [comment]
> In theory we have the machinery to do it without _linking_ the library. However it looks like ELF::DynamicObject expects that the library is created from the DynamicLoader::map method, which does require at least loading the program headers (and all loadable sections :( ) somewhere in memory. Which is the solution presented in this PR.
> 
> I think we could definitely do less work with some LibELF re-architecting though. We shouldn't have to actually mmap all of the loadable segments into memory in order to check the dynamic section for DT_NEEDED entries.
> 
> We would just need to, for each library:
> 
>     * Open it up and find the dynamic section
> 
>     * Grab all the DT_NEEDED entries
> 
>     * Looking at LD_LIBRARY_PATH, builtin library paths, and RPATH of the current object, find the path to all the DT_NEEDED objects
> 
>     * Keep doing that recursively.
> 
> 
> Doing that in a way that will stay in lock step with the way dependencies are resolved in the loader itself sounds hard though. Any differences in how they resolve libraries could cause some serious headaches when checking ldd vs running a binary.

I see your point and I agree that less work can be done. Could we do that in a future PR though? I'd like to do incrementally and keep the pull request patches as small as possible for now.

File: Userland/Utilities/ldd.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+{
+    if (recusrive_iteration > recursive_iteration_max)
+        return ELOOP;
+
+    Vector<DeprecatedString> libraries;
+    object.for_each_needed_library([&libraries](StringView entry) {
+        libraries.append(DeprecatedString::formatted("{}", entry));
+    });
+    for (auto& library_name : libraries) {
+        auto possible_library_path = ELF::DynamicLinker::resolve_library(library_name, object);
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/ldd.cpp:38]: @ADKaster I don't see how we could drift away from LibELF resolution mechanism given that we use this function here and could also use it in a non-loading resolver?

File: Userland/Utilities/ldd.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            continue;
+        }
+
+        RefPtr<ELF::DynamicObject> library_object = loader->map();
+        if (!library_object) {
+            outln("Failed to map dynamic ELF object {}", library_path);
+            continue;
+        }
+        outln("{} => {}", library_name, library_path);
+        recusrive_iteration++;
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/ldd.cpp:75]: Typo

File: Userland/Utilities/ldd.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            outln("{} is not a valid ELF dynamic shared object!", path);
+            return 1;
+        }
+
+        object = loader->map();
+        if (!object) {
+            outln("Failed to map dynamic ELF object {}", path);
+            return 1;
+        }
+        if (recursive_iteration_max.has_value())
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/ldd.cpp:145]: You can use value_or on the recursion depth to avoid duplicating the call

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
I don't understand ELFs but this looks nice! The Andrew ping is not relevant for the review.

File: Userland/Utilities/ldd.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#include <LibELF/DynamicObject.h>
+#include <LibELF/Image.h>
+#include <LibELF/Validation.h>
+#include <LibMain/Main.h>
+#include <ctype.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <unistd.h>
+
+static Vector<DeprecatedString> found_libraries;
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/Utilities/ldd.cpp:26]: I tried to do that and it involves changing too many unrelated things in the codebase to do that. I rather send it as a separate PR.

**Comment by mistafunk** [comment]
FWIW, had a quick play with this PR while trying to get OpenJDK going and noticed that the value of `interpreter_path_builder` is mostly empty (did not found a pattern yet)... specifying the `-f` flag works though.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

**Pull request merged**